cpp/src/arrow/flight/transport/ucx/ucx_internal.cc

```diff
@@ -0,0 +1,1164 @@
+    if (status == UCS_OK) {
+      pending_send->completed.MarkFinished();
+    } else {
+      pending_send->completed.MarkFinished(FromUcsStatus("ucp_am_send_nbx", status));
+    }
+    // TODO(lidavidm): delete should occur on a background thread if there's mapped
+    // buffers, since unmapping can be nontrivial and we don't want to block
+    // the thread doing UCX work. (Borrow the Rust transfer-and-drop pattern.)
+    delete pending_send;
+    ucp_request_free(request);
+  }
+
+  static void AmRecvCallback(void* request, ucs_status_t status, size_t length,
+                             void* user_data) {
+    auto* pending_recv = reinterpret_cast<PendingAmRecv*>(user_data);
+    ucp_request_free(request);
+    if (status != UCS_OK) {
+      pending_recv->driver->Push(
+          FromUcsStatus("ucp_am_recv_data_nbx (callback)", status));
+    } else {
+      pending_recv->driver->Push(std::move(pending_recv->frame));
+    }
+    delete pending_recv;
+  }
+
+  arrow::Result<ucs_status_t> RecvActiveMessageImpl(const void* header,
+                                                    size_t header_length, void* data,
+                                                    const size_t data_length,
+                                                    const ucp_am_recv_param_t* param) {
+    DCHECK(param->recv_attr & UCP_AM_RECV_ATTR_FIELD_REPLY_EP);
+
+    if (data_length > static_cast<size_t>(std::numeric_limits<int64_t>::max())) {
+      return Status::Invalid(
+          "Cannot allocate buffer greater than int64_t max, requested: ", data_length);
+    }
+
+    ARROW_ASSIGN_OR_RAISE(auto frame, Frame::ParseHeader(header, header_length));
+    if (data_length < frame->size) {
+      return Status::IOError("Expected frame of ", frame->size, " bytes, but got only ",
+                             data_length);
+    }
+
+    if ((param->recv_attr & UCP_AM_RECV_ATTR_FLAG_DATA) &&
+        (frame->type != FrameType::kPayloadBody || memory_manager_->is_cpu())) {
+      // Zero-copy path. UCX-allocated buffer must be freed later.
+
+      // XXX: this buffer can NOT be freed until AFTER we return from
+      // this handler. Otherwise, UCX won't have fully set up its
+      // internal data structures (allocated just before the buffer)
+      // and we'll crash when we free the buffer. Effectively: we can
+      // never use Then/AddCallback on a Future<> from ReadFrameAsync,
+      // because we might run the callback synchronously (which might
+      // free the buffer) when we call Push here.
+      frame->buffer =
+          arrow::internal::make_unique<UcxDataBuffer>(worker_, data, data_length);
+      Push(std::move(frame));
+      return UCS_INPROGRESS;
+    }
+
+    if ((param->recv_attr & UCP_AM_RECV_ATTR_FLAG_DATA) ||
+        (param->recv_attr & UCP_AM_RECV_ATTR_FLAG_RNDV)) {
+      // Rendezvous protocol (RNDV), or unpack to destination (DATA).
+
+      // We want to map/pin/register the buffer for faster transfer
```

Comment:
This all runs on the CPU. This will (depending on the transport) result in the buffer being page-locked so that network devices/GPU can access it via DMA without having to worry about the memory being paged out. I found that on Infiniband this makes a big difference in throughput.